Feat: header, option button 동적 UI 구현

Pangho04 opened the pull request:
<!---- <타입>: <일감 명> 의 형식으로 제목 작성 -->

## 개요

<!---- 변경 사항 및 관련 이슈가 있다면 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
<!---- Resolves: #15 -->
### [Header]
- [x]  입력창 위에 현재 보여주고 있는 아티클 카드의 리딩 타임 합계를 표시합니다.
    - [x]  저장된 데이터가 없는 경우<InfoConcepText> 메세지를 표시해 줍니다.
- [x]  사용자가 URL을 입력하고 엔터를 누르거나 ‘붙여넣기’시 API를 요청합니다.
- [x] 한 번에 최대 30개의 url만 입력이 가능합니다.(유효성)
- [x] 서버 요청 시 전달하는 정보는 다음과 같습니다.
```
GET /articleSummary

request parameter
  url(required): url 주소
  wpm(required): 1분당 읽는 단어 수
```
- [x] 반환값을 로컬 스토리지에 저장합니다.
```

  {
	  "createDate": "2024-06-13T06:17:25.710Z",
	  data: {
	    "readingTime": 600000,
	    "title": "JavaScript",
	    "siteName": "MDN",
	    "url": "https://developer.mozilla.org",
	    "faviconUrl": "https://developer.mozilla.org/favicon-48x48.cbbd161b.png"
	  }
	  statusCode: 200,
  }
```
- [x] 새로운 데이터가 목록에 포함되어 나타납니다.
- [x] 새로운 데이터가 목록에 포함되어 나타납니다.
- [x] 유효하지 않은 URL 형식일 경우 토스트 컴포넌트를 보여줘야 합니다.
- [x] API 요청의 응답 상태 코드가 200이 아닐 경우 토스트 컴포넌트를 보여줘야 합니다. 
- [x] 아티클을 저장한 경우 아티클의 개수를 확인하여 30의 배수가 될 경우 토스트 컴포넌트를 보여줘야 합니다. 
### [Option]
- [x]  옵션 버튼 클릭 시 ‘재설정 버튼, 수정 버튼’이 표시되어야 합니다.
    - 표시된 ‘재설정 버튼, 수정 버튼’을 클릭하지 않는 경우
        - [x]  재설정 버튼을 재클릭 할 경우 표시된 버튼이 사라집니다.
        - [x]  버튼이 아닌 다른 곳을 클릭 할 경우 표시된 버튼이 사라집니다.
    - 재설정 버튼을 클릭하는 경우
        - [x]  모달 3으로 이동합니다.
    - 수정 버튼을 클릭하는 경우
        - [x]  표시된 버튼이 사라집니다.
        - [x]  아티클 카드에 삭제 버튼이 표시됩니다.

## 코드 변경 사항

### useState()
https://github.com/team-sticky-252/readim-client/blob/cb5b2902db90deacd03c17b519ea00eba601d4ac/src/App.jsx#L42-L53
최상단 컴포넌트 App에서 State 생성 후, props 전달로 정보 흐름 관리.
위에서 언급한 각종 기능 구현.
### Promise.all
https://github.com/team-sticky-252/readim-client/blob/cb5b2902db90deacd03c17b519ea00eba601d4ac/src/components/Header/UrlInputContainer.jsx#L72-L81
fetch 비동기 병렬 요청 및 로컬 스토리지 저장을 위해 promis.all 사용.
### message.json, urlUtils.js
https://github.com/team-sticky-252/readim-client/blob/cb5b2902db90deacd03c17b519ea00eba601d4ac/src/utils/message.json#L1-L10
상수 파일 별도 생성.

## 기타 전달 사항

<!---- 리뷰어들에게 기능관련 외에 전달하고 싶은 사항을 작성해주세요. -->

## PR Checklist

<!---- PR이 다음 요구 사항을 충족하는지 확인하세요. -->

- [x] 주어진 Task의 요구사항을 모두 작성했습니다.
- [x] 작성한 PR을 다시 한번 더 검토하였습니다.
- [x] merge 할 브랜치의 위치를 확인하였습니다.

Pangho04 committed:
Design: ConceptTextBox CSS 수정

Pangho04 committed:
Feat: 기능별 State 생성

-<HeaderContainer>
  1. LocalStorage 저장값 갱신 State
  2. ErrorMessage 저장값 갱신 State
  3. 편집 모드 여부 State

Pangho04 committed:
Feat: 총 ReadTime 안내 기능

- 로컬 스토리지 상의 정보 업데이트시, 총 리드 타임 갱신 후 표시.
- 총 리드 타임 0일 시, 컨셉 텍스트 표시.

Pangho04 committed:
Feat: 옵션 버튼 동적 UI 기능 구현

- 옵션 버튼 클릭 시 팝업 버튼 표시.(기본: 숨김)
- 팝업 버튼 바깥 클릭 시 팝업 버튼 숨김.
- 팝업 버튼 > edit 클릭 시, 카드 삭제 버튼 표시.

Pangho04 committed:
Feat: URL 입력 창 동적 UI 구현

- URL 최대 30개 까지 입력 가능.
- 엔터 버튼 키다운or 붙여넣기 시 request 요청.
- response 유효성 검사 후, 로컬 스토리지 저장.
- 잘못된 입력값, , 중복값, response 상대값 등 오류별 toast 메시지 표시.

Pangho04 committed:
Rename: 상수 토스트 메세지 파일로 수정

- 기존 toast 메세지 목업 파일 -> toast 메세지 상수 파일로 수정

Pangho04 committed:
Add: URL 입력 관련 함수 Util 파일

- <UrlInputContainer>의 request 요청, 유효성 검사 등 함수 외부화.

Pangho04 committed:
Style: 사전 merge건 rebase중 수정

Pangho04 committed:
Style: Lint 오류로 인한 수정

JieunZoeyOh commented on src/components/Header/UrlInputContainer.jsx:
```diff
@@ -1,29 +1,115 @@
+          setMessageList,
+          textarea,
+          message,
+          articleDatas.data ? articleDatas.data.url : null,
```
```suggestion
          articleDatas.data?.url,
```

이렇게 쓸 수 있을 것 같습니다.

Pangho04 replied:
해당 사항 수정했습니다!

dlgnswk commented on src/App.jsx:
```diff
@@ -39,11 +39,28 @@ function App() {
 
+  const [messageList, setMessageList] = useState([]);
+  const [isDeleteMode, setIsDeleteMode] = useState(false);
+  const [prevArticleDatas, setPrevArticleDatas] = useState([]);
```
```suggestion
  const [messageList, setMessageList] = useState([]);
  const [isDeleteMode, setIsDeleteMode] = useState(false);
  const [prevArticleDataList, setPrevArticleDataList] = useState([]);
```
messageList 는 메세지들을 담는 리스트라는 뜻이라고 이해됩니다!
일관성을 위해 아티클 데이터를 담는 `prevArticleDatas` 또한 `prevArticleDataList`로 변경하는거나 `messageList`를 `messages` 로 변경하는 것은 어떨까요?

Pangho04 replied:
prevArticleDatas => prevArticleDataList
setPrevArticleDatas => setPrevArticleDataList
로 변경했습니다

dlgnswk commented on src/App.jsx:
```diff
@@ -39,11 +39,28 @@ function App() {
+    const storedURLs = JSON.parse(window.localStorage.getItem("URLs"));
+
+    if (storedURLs) {
+      setPrevArticleDatas(storedURLs);
```
47번 라인에서는 로컬 스토리지에서 키 `URLs`를 통해 `storedURLs` 변수에 url들을 받아온다고 이해됩니다.
50번 라인에서는 url들을 아티클 데이터에 담는다는 것이 조금 애매한것 같습니다!
`articleDatas` 또는 `articleDataList`는 어떨까요?

Pangho04 replied:
URLs 이란 키 값에 담김 url 정보들을 불러오는 것 이므로, storedURLs가 더 적합해 보입니다.
 또한, 해당 값이 prevArticleDataList와 동일한 내용이므로, articleDataList로 설정할 경우 겹치게 된다고 생각합니다.
 또, 해당 변수명을 하위 컴포넌트에서 fetch를 통해 불러온 current 값을 담는 변수명으로서 사용하고 있으므로, 해당 변경은 지양하고자 합니다.

JieunZoeyOh commented on src/utils/urlUtils.js:
```diff
@@ -0,0 +1,125 @@
+) => {
+  const numOfURL = inputValue.split("http").length - 1;
+
+  if (numOfURL > 30) {
```
isValid 함수가 사용되는 곳인 UrlInputContainer.jsx를 확인해보니 이 조건문이 true인 케이스는 안나올 것 같아요. 확인 부탁드립니다!

Pangho04 replied:
해당 유효성 검사 로직 UrlInputContainer 컴포넌트로 이동 했습니다

JieunZoeyOh commented on src/utils/urlUtils.js:
```diff
@@ -0,0 +1,125 @@
+      {
+        id: prev ? prev.length + 1 : 1,
+        message: message[0].onlySpace,
+        link: link || null,
```
```suggestion
        link,
```

위(34번째 라인)에서 default 값(null)을 주고 있어서 생략할 수 있을 것 같아요.

Pangho04 replied:
해당 줄 포함, 다른 로직들에도 반영했습니다.

JieunZoeyOh commented on src/components/Header/UrlInputContainer.jsx:
```diff
@@ -1,29 +1,115 @@
-    textarea.current.style.height = `${newHeight}px`;
+        articleDatas.createDate = new Date(
+          new Date().toLocaleString("en-US", { timeZone: "Asia/Seoul" }),
+        ).toISOString();
```
```suggestion
        articleDatas.createDate = new Date().toISOString();
```

로 하셔도 될 것 같습니다.

Pangho04 replied:
제안하신 값을 한번 더 new Date()로 감싸시는 걸 깜빡하신 것 같아서, 한번 더 감싸 수정했습니다.

dlgnswk reviewed (changes requested):
코멘트 확인 부탁드립니다!

JieunZoeyOh reviewed (changes requested):
고생하셨습니다. 리뷰 확인 한 번만 부탁드릴게요.

Pangho04 committed:
Style: State명 수정

- prevArticleDatas => prevArticleDataList
-setPrevArticleDatas => setPrevArticleDataList

Pangho04 committed:
Style: 로직 변경으로 인한 구조 수정

- url fetch 함수 파일 urlUtiles 유효성 검사 로직 이동.
- 기본값 설정으로 필요 없어진 조건문 삭제.

Pangho04 reviewed (commented):
확인 부탁드립니다~

Pangho04 committed:
Style: 코드 스타일 수정

Pangho04 committed:
Style: return문 개행 통일
